tc-server/src/main/java/com/tc/net/groups/TCGroupManagerImpl.java

```diff
@@ -1076,37 +1078,27 @@ protected void switchToState(HandshakeState state) {
         debugInfo("[TCGroupHandshakeStateMachine]: Attempting to switch state (" + current + "->" + state + "): "
                 + stateInfo(state));
       }
+      HandshakeMonitor previous = null;
+      HandshakeMonitor next = state.createMonitor();
+      
       synchronized (this) {
-        if (current == STATE_FAILURE) {
+        previous = this.current;
+        if (current.getState() == STATE_FAILURE) {
           if (isDebugLogging()) {
             debugWarn("Ignored switching to " + state + " as current is " + current + ", " + stateInfo(state));
           }
           return;
         }
-        this.current = state;
-        waitForStateTransitionToComplete();
-        stateTransitionInProgress = true;
+        this.current = next;
       }
+      
       if (isDebugLogging()) {
         debugInfo("[TCGroupHandshakeStateMachine]: Entering state: " + state + ", for channel: " + channel);
       }
+      previous.waitForCompletion();
+      next.start();
```

Comment:
Could we embed the `state.enter()` call inside the monitor and make it:

```
previous.waitForComplete()
next.enter()
```

Where `next.enter()` would "lock", call and then "unlock"?

Reply:
good suggestion.